templates/_helpers.tpl

```diff
@@ -0,0 +1,195 @@
+{{- default .Chart.Name .Values.nameOverride | trunc 63 | trimSuffix "-" }}
+{{- end }}
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+If release name contains chart name it will be used as a full name.
+*/}}
+{{- define "tonic.fullname" -}}
+{{- if .Values.fullnameOverride }}
+{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
+{{- else }}
+{{- $name := default .Chart.Name .Values.nameOverride }}
+{{- if contains $name .Release.Name }}
+{{- .Release.Name | trunc 63 | trimSuffix "-" }}
+{{- else }}
+{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" }}
+{{- end }}
+{{- end }}
+{{- end }}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "tonic.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" }}
+{{- end }}
+
+{{/*
+Common labels
+*/}}
+{{- define "tonic.labels" -}}
+helm.sh/chart: {{ include "tonic.chart" . }}
+{{ include "tonic.selectorLabels" . }}
+{{- if .Chart.AppVersion }}
+app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
+{{- end }}
+app.kubernetes.io/managed-by: {{ .Release.Service }}
+{{- end }}
+
+{{/*
+Selector labels
+*/}}
+{{- define "tonic.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "tonic.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end }}
+
+{{/*
+Create the name of the service account to use
+*/}}
+{{- define "tonic.serviceAccountName" -}}
+{{- if .Values.serviceAccount }}
+{{- if .Values.serviceAccount.create -}}
+    {{ default (include "tonic.fullname" .) .Values.serviceAccount.name }}
+{{- else -}}
+    {{ default "default" .Values.serviceAccount.name }}
+{{- end -}}
+{{- else -}}
+    {{ "" }}
+{{- end -}}
+{{- end }}
+
+{{- define "tonic.hostIntegration" -}}
+{{- if ((.Values.tonicai).web_server).features -}}
+{{- .Values.tonicai.web_server.features.host_integration_enabled -}}
+{{- else -}}
+{{ "false" }}
+{{- end -}}
+{{- end }}
+
+{{/*
+Given the $.tonicSsoConfig, create all environment variables needed for
+the deployment. This should only be called if $.tonicSsoConfig is populated
+*/}}
+{{- define "tonic.sso" -}}
+{{- $provider := (required "tonicSsoConfig.provider is required for SSO" .provider | lower) -}}
+- name: TONIC_SSO_PROVIDER
+  value: {{ quote .provider }}
+{{- if .groupFilter -}}
+- name: TONIC_SSO_GROUP_FILTER_REGEX
+  value: {{quote .groupFilter}}
+{{- end -}}
+{{- if eq $provider "aws" }}
+{{ include "tonic.sso.aws" . }}
+{{- else if eq $provider "azure" }}
+{{ include "tonic.sso.azure" . }}
+{{- else if eq $provider "duo" }}
+{{ include "tonic.sso.duo" . }}
+{{- else if eq $provider "google" }}
+{{ include "tonic.sso.google" . }}
+{{- else if eq $provider "okta" }}
+{{ include "tonic.sso.okta" . }}
+{{- else if eq $provider "keycloak" }}
+{{ include "tonic.sso.keycloak" . }}
+{{- else if eq $provider "saml" }}
```

Comment:
This is some next level validation lol

Reply:
Easier to keep things required for each provider together in their own named template.